pkg/controller/windowsmachine/windowsmachine_controller.go

```diff
@@ -264,60 +265,87 @@ func (r *ReconcileWindowsMachine) Reconcile(request reconcile.Request) (reconcil
-					// Delete already initiated
+			node := &core.Node{}
+			err := r.client.Get(context.TODO(), kubeTypes.NamespacedName{Namespace: machine.Status.NodeRef.Namespace,
+				Name: machine.Status.NodeRef.Name}, node)
+			if err != nil {
+				return reconcile.Result{}, errors.Wrapf(err, "could not get node associated with machine %s", machine.GetName())
+			}
+			if _, present := node.Annotations[nodeconfig.VersionAnnotation]; present {
+				// version annotation doesn't match the current operator version
+				if node.Annotations[nodeconfig.VersionAnnotation] != version.Get() {
+					machinesetName := "unknown"
+					if len(machine.OwnerReferences) > 0 {
+						machinesetName = machine.OwnerReferences[0].Name
+					}
+					log.Info("upgrading machineset", "name", machinesetName)
+					if !r.isAllowedDeletion(machine) {
+						log.Info("machine deletion restricted", "name", machine.GetName(),
+							"maxUnhealthyCount", maxUnhealthyCount)
+						r.recorder.Eventf(machine, core.EventTypeWarning, "MachineDeletionRestricted",
+							"Machine %v deletion restricted as the maximum unhealthy machines can`t exceed %v count",
+							machine.Name, maxUnhealthyCount)
+						return reconcile.Result{Requeue: true}, nil
+					}
+					if !machine.GetDeletionTimestamp().IsZero() {
+						// Delete already initiated
+						return reconcile.Result{}, nil
+					}
+
+					if err := r.client.Delete(context.TODO(), machine); err != nil {
+						r.recorder.Eventf(machine, core.EventTypeWarning, "MachineDeletionFailed",
+							"Machine %v deletion failed: %v", machine.Name, err)
+						return reconcile.Result{}, err
+					}
+					log.Info("machine has been remediated by deletion", "name", machine.GetName())
+					r.recorder.Eventf(machine, core.EventTypeNormal, "MachineDeleted",
+						"Machine %v has been remediated by deleting the Machine object", machine.Name)
 					return reconcile.Result{}, nil
 				}
+				log.Info("machine has current version", "name", machine.GetName(),
+					"version", node.Annotations[nodeconfig.VersionAnnotation])
+				// version annotation exists with a valid value, node is fully configured, do nothing.
+				// Reconfigure prometheus if endpoints object does not have entry for current node.
+				if err := pc.Configure(true); err != nil {
+					return reconcile.Result{}, errors.Wrapf(err, "failed to configure Prometheus for node %s", node.Name)
 
-				if err := r.client.Delete(context.TODO(), machine); err != nil {
-					r.recorder.Eventf(machine, core.EventTypeWarning, "MachineDeletionFailed",
-						"Machine %v deletion failed: %v", machine.Name, err)
-					return reconcile.Result{}, err
 				}
-				log.Info("machine has been remediated by deletion", "name", machine.GetName())
-				r.recorder.Eventf(machine, core.EventTypeNormal, "MachineDeleted",
-					"Machine %v has been remediated by deleting the Machine object", machine.Name)
+				log.Info("prometheus is reconfigured for the node", "name", node.Name)
```

Comment:
Please move this log message to spot where the actual reconfiguration takes place. Otherwise this message will always appear even if we are not updating the endpoint object.